app/server/models/credit_transfer.py

```diff
@@ -84,23 +86,87 @@ def transfer_amount(self, val):
     def send_blockchain_payload_to_worker(self, is_retry=False, queue='high-priority'):
         sender_approval = self.sender_transfer_account.get_or_create_system_transfer_approval()
         recipient_approval = self.recipient_transfer_account.get_or_create_system_transfer_approval()
+
+        approval_priors = list(
+            filter(lambda x: x is not None,
+                   [
+                       sender_approval.eth_send_task_uuid, sender_approval.approval_task_uuid,
+                       recipient_approval.eth_send_task_uuid, recipient_approval.approval_task_uuid
+                   ]))
+
+        other_priors = [t.blockchain_task_uuid for t in self._get_required_prior_tasks()]
+
+        all_priors = approval_priors + other_priors
+
         return bt.make_token_transfer(
             signing_address=self.sender_transfer_account.organisation.system_blockchain_address,
             token=self.token,
             from_address=self.sender_transfer_account.blockchain_address,
             to_address=self.recipient_transfer_account.blockchain_address,
             amount=self.transfer_amount,
-            prior_tasks=
-            list(filter(lambda x: x is not None,
-                        [
-                            sender_approval.eth_send_task_uuid, sender_approval.approval_task_uuid,
-                            recipient_approval.eth_send_task_uuid, recipient_approval.approval_task_uuid
-                        ])),
+            prior_tasks=all_priors,
             queue=queue,
             task_uuid=self.blockchain_task_uuid
         )
 
-    def resolve_as_completed(self, existing_blockchain_txn=None, queue='high-priority'):
+    def _get_required_prior_tasks(self):
+        """
+        Get the tasks involving the sender's account that must complete prior to this task being submitted to chain
+        To calculate the prior tasks for the sender Alice:
+
+        - Find the most recent credit transfer where Alice was the sender, not including any transfers that have the
+            same batch UUID as this transfer. Call this "most_recent_out_of_batch_send"
+        - Find all credit transfers subsequent to "most_recent_out_of_batch_send" where Alice was the recipient. Call
+            this "more_recent_receives"
+
+        Required priors are all transfers in "more_recent_receives" and "most_recent_out_of_batch_send".
+        For why this works, see https://github.com/teamsempo/SempoBlockchain/pull/262
+
+        """
+
+        # We're constantly querying complete transfers here. Lazy and DRY
+        complete_transfer_base_query = (
+            CreditTransfer.query.filter(CreditTransfer.transfer_status == TransferStatusEnum.COMPLETE)
```

Comment:
Could we also filter `BlockchainStatus == SUCCESS` here? Otherwise (if I'm reading this correctly), we'd be sending a bunch of priors for tasks we know for sure are already complete from a tx callback.

Reply:
Yeah that's true, but I also do like the idea of having the complete dependency graph written down somewhere.
Otoh it does use up data.

Reply:
It could be a separate function to get the complete dependency graph if we need it. My thinking around sending one with `BlockchainStatus == SUCCESS` filtered is that, IMO, we should only send the minimum amount of information the worker needs to complete the task. I think sending less data with this would make the task dependencies more readable and easier to debug, be more efficient, and prevent performance bloat and slow-down on long running instances. An extreme example would be a retail store having to check that a candy somebody bought a few years ago successfully cleared.

Reply:
Fair point! added here: https://github.com/teamsempo/SempoBlockchain/pull/262/files#diff-a140437659d7c51f9bfbfc2afbf84ad1R208